server/src/modules/users/domain/entities/auth-secret/auth-secret.ts

```diff
@@ -0,0 +1,34 @@
+import { Result } from '../../../../../shared/core/result'
+import { AggregateRoot } from '../../../../../shared/domain/aggregate-root'
+import { UniqueEntityID } from '../../../../../shared/domain/unique-entity-id'
+import { EncryptedClientSecret } from '../../value-objects/encrypted-client-secret'
+
+interface AuthSecretProps {
+  clientId: string,
+  encryptedClientSecret: EncryptedClientSecret
+}
+
+export class AuthSecret extends AggregateRoot<AuthSecretProps> {
```

Comment:
Would we not also want to log the creation of an AuthSecret through a DomainEvent?

Reply:
My motivation behind that was we don't actually populate the `AuthSecret` collection through an endpoint dynamically. We would add them as needed using an external script (and thus wouldn't need a `DomainEvent` at all for security reasons).

Reply:
Could you elaborate on how we plan on populating the `AuthSecret` collection using a script?

Reply:
The idea is that we want to be very selective about how we populate the `AuthSecret` collection; when our microservice administers the `clientId` and `secret` to service providers we are beg. 

Since we're only going to have **at maximum one user** of this microservice in the next few months, I don't think it's worth making something like a service registration UI where the service provider is provided an ID and secret. 

Alternatively, we could make a `POST` endpoint that does the following: 

1. Creates an `AuthSecret` entity and saves it, with an additional flag `is_verified` to false.
2. Responds the newly created `client_id` and `secret`.

Now the service provider would save these values locally (securely) and use them when redirecting to the microservice login page, and using the token endpoint. The manager/developer of the service provider would contact us with their `client_id` and we would switch the `is_verified` tag on the relevant instance of the microservice to True, using a DB script, enabling the aforementioned methods. I don't believe there's anything inherently unsafe about this method but will continue to research this. This is similar to how apps like Github administer ids and secrets:
[](https://www.oauth.com/oauth2-servers/client-registration/client-id-secret/)

I will capture this in a Zenhub ticket. Note that this requires generating a `clientId` dynamically based on the given client name and some factor like the time of creation (something that isn't currently done with the `clientId`; it's simply a raw string that doesn't have the same encapsulation as the `EncryptedClientSecret`). Perhaps we can include the necessary refactor in that ticket and leave it out of this PR.